### PR TITLE
JS/Ruby: Represent non-positional arguments with Argument/Parameter tokens

### PR DESCRIPTION
**Pull request by asgerf**
Previously we had separate tokens for non-positional arguments, like `Receiver` and `BlockArgument`.

Now these are represented as operands to `Argument`/`Parameter` tokens, as discussed with @hvitved:

Ruby:
- `Argument[self]` for the receiver (previously `Receiver`)
- `Argument[block]` for the block argument (previously `BlockArgument`)
- `Argument[foo:]` for a keyword argument named `foo` (previously not implemented)

JS:
- `Argument[this]` for the receiver (previously `Argument[-1]`)

All of the above also apply to the `Parameter` token.

The names `self` and `this` were chosen to coincide with the keyword/convention from the language.

### CODE REVIEW

**Review by hmac — APPROVED** [review]
I don't feel qualified to comment closely on the implementation, but it seems straightforward and the test changes are what I'd expect. I think I have a slight preference for `Receiver` over `Argument[self]` (shorter and matches better with `MethodCall.getReceiver`) but having consistency between the languages probably trumps that.

The keyword parameter support is really nice!

**Review by erik-krogh — DISMISSED** [review]
JS :+1:  
(The JS part is simple).

File: ruby/ql/lib/codeql/ruby/ApiGraphs.qll [hunk 1]
```diff
@@ -803,5 +773,37 @@ module API {
...
+        result = Label::parameter(n)
+      )
+      or
+      exists(string name |
+        pos.isKeyword(name) and
+        result = Label::keywordParameter(name)
+      )
+      or
+      pos.isBlock() and
+      result = Label::blockParameter()
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/ApiGraphs.qll:790]: Should the `toString` of `LabelBlockParameter` be updated to `getParameter(block)` (and similar for `LabelKeywordParameter`? And are we missing labels for the `self` parameter?

File: ruby/ql/test/library-tests/dataflow/summaries/summaries.rb [hunk 2]
```diff
@@ -47,3 +47,20 @@ def userDefinedFunction(x, y)
...
 end
+
+sink(Foo.namedArg(foo: tainted))
+sink(Foo.namedArg(tainted))
+
+Foo.intoNamedCallback(tainted, foo: ->(x) {
+  sink(x)
+})
+Foo.intoNamedParameter(tainted, ->(foo:) {
+  sink(foo)
```
Comment (hvitved) [inline ruby/ql/test/library-tests/dataflow/summaries/summaries.rb:58]: We should document that flow is missing in this case

**Review by hvitved — COMMENTED** [review]
LGTM; two minor comments.

File: ruby/ql/lib/codeql/ruby/ApiGraphs.qll (hunk 1 above)
Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/ApiGraphs.qll:790]: The `toString` value of these labels are currently named after predicates on `API::Node` (not MaD tokens). At least that seemed to be the convention, one that I've tried to continue. So it doesn't need updating.

File: ruby/ql/test/library-tests/dataflow/summaries/summaries.rb (hunk 2 above)
Comment (asgerf) [reply ruby/ql/test/library-tests/dataflow/summaries/summaries.rb:58]: Dang, I missed that. I'll try to introduce inline assertions to this test.

File: ruby/ql/lib/codeql/ruby/ApiGraphs.qll (hunk 1 above)
Comment (erik-krogh) [reply ruby/ql/lib/codeql/ruby/ApiGraphs.qll:790]: I like having the `toString()` be the associated predicate on `API::Node`, so maybe we should update JS to also do that?

File: ruby/ql/test/library-tests/dataflow/summaries/summaries.rb (hunk 2 above)
Comment (asgerf) [reply ruby/ql/test/library-tests/dataflow/summaries/summaries.rb]: Updated to use inline test expectations, much better overall, and it's now clear that this flow is missing.

I haven't investigated why the flow is missing, though. I was under the impression that providing the correct `ArgumentPosition` and `ParameterPosition` values would be enough.

**Review by erik-krogh — DISMISSED** [review]

File: ruby/ql/test/library-tests/dataflow/summaries/summaries.rb (hunk 2 above)
Comment (asgerf) [reply ruby/ql/test/library-tests/dataflow/summaries/summaries.rb]: @hvitved would you like it addressed in this PR? Please let me know either way

Comment (hvitved) [reply ruby/ql/test/library-tests/dataflow/summaries/summaries.rb]: No, let's do that follow-up

File: ruby/ql/test/library-tests/dataflow/summaries/Summaries.ql [hunk 3]
```diff
@@ -117,23 +117,18 @@ private class SinkFromModel extends ModelInput::SinkModelCsv {
...
-    )
+    super.isSink(sink)
     or
     sink = ModelOutput::getASinkNode("test-sink").getARhs()
   }
 }
-
-from DataFlow::PathNode source, DataFlow::PathNode sink, Conf conf
-where conf.hasFlowPath(source, sink)
-select sink, source, sink, "$@", source, source.toString()
```
Comment (hvitved) [inline ruby/ql/test/library-tests/dataflow/summaries/Summaries.ql:134]: I prefer to actually keep this, in addition to `import PathGraph`, like we do in other flow tests that use `InlineFlowTest`.

**Review by hvitved — APPROVED** [review]

**Review by erik-krogh — APPROVED** [review]

**Pull request merged**